capabilityquerier/pkg/capabilityquerier/capabilityquerier.go

```diff
@@ -0,0 +1,189 @@
+package capabilityquerier
+
+import (
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/onc-healthit/lantern-back-end/lanternmq"
+	"github.com/pkg/errors"
+)
+
+var fhir3PlusJSONMIMEType = "application/fhir+json"
+var fhir2LessJSONMIMEType = "application/json+fhir"
+
+var ssl30 = "SSL 3.0"
+var tls10 = "TLS 1.0"
+var tls11 = "TLS 1.1"
+var tls12 = "TLS 1.2"
+var tls13 = "TLS 1.3"
+var tlsUnknown = "TLS version unknown"
+
+type Message struct {
+	URL                 string      `json:"url"`
+	Err                 string      `json:"err"`
+	MimeType            string      `json:"mimetype"`
+	TLSVersion          string      `json:"tlsVersion"`
+	CapabilityStatement interface{} `json:"capabilityStatement"`
+}
+
+func GetAndSendCapabilityStatement(
+	ctx context.Context,
+	fhirURL *url.URL,
+	client *http.Client,
+	mq *lanternmq.MessageQueue,
+	ch *lanternmq.ChannelID,
+	queueName string) error {
+
+	var err error
+	message := Message{
+		URL: fhirURL.String(),
+	}
+
+	capResp, mimeType, tlsVersion, err := requestCapabilityStatement(ctx, fhirURL, client)
+	if err == nil {
+		message.MimeType = mimeType
+		message.TLSVersion = tlsVersion
+		err = json.Unmarshal(capResp, &(message.CapabilityStatement))
+		if err != nil {
+			message.Err = err.Error()
+		}
+	} else {
+		message.Err = err.Error()
+	}
+
+	msgBytes, err := json.Marshal(message)
+	if err != nil {
+		return errors.Wrapf(err, "error marshalling json message for request to %s", fhirURL.String())
+	}
+	msgStr := string(msgBytes)
+
+	err = sendToQueue(ctx, msgStr, mq, ch, queueName)
+	if err != nil {
+		return errors.Wrapf(err, "error sending capability statement to queue '%s' over channel '%v'", queueName, *ch)
+	}
+
+	return nil
+}
+
+func requestCapabilityStatement(ctx context.Context, fhirURL *url.URL, client *http.Client) ([]byte, string, string, error) {
+	var err error
+	var resp *http.Response
+
+	req, err := http.NewRequest("GET", fhirURL.String(), nil)
+	if err != nil {
+		return nil, "", "", errors.Wrap(err, "unable to create new GET request from URL: "+fhirURL.String())
+	}
+	req = req.WithContext(ctx)
+
+	// make the request using a JSON mime type to get a JSON response.
+	// track the mime type to provide to the queue as data re the request.
+	tryOtherMimeType := false
+	mimeType := fhir3PlusJSONMIMEType
+	resp, err = requestWithMimeType(req, mimeType, client)
+	if err != nil {
+		// it's possible that the error is due to a 406 Not Acceptable error for the wrong meme type, so we
```

Comment:
~meme~ mime

Also, should we check for a 406 specifically that way we are not reissuing the request with a different mime-type in the event of 404s,500s etc...?

Reply:
addressed